File: app/src/main/java/de/tum/in/tumcampusapp/activities/PersonsSearchActivity.java
package de.tum.in.tumcampusapp.activities;

import android.content.Intent;
import android.database.Cursor;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.ArrayAdapter;
import android.widget.ListAdapter;
import android.widget.ListView;

import java.util.ArrayList;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.activities.generic.ActivityForSearchingTumOnline;
import de.tum.in.tumcampusapp.adapters.NoResultsAdapter;
import de.tum.in.tumcampusapp.auxiliary.PersonSearchSuggestionProvider;
import de.tum.in.tumcampusapp.managers.RecentsManager;
import de.tum.in.tumcampusapp.models.tumo.Person;
import de.tum.in.tumcampusapp.models.tumo.PersonList;
import de.tum.in.tumcampusapp.tumonline.TUMOnlineConst;

/**
 * Activity to search for employees.
 */
public class PersonsSearchActivity extends ActivityForSearchingTumOnline<PersonList> implements OnItemClickListener {
    private static final String P_SUCHE = "pSuche";

    /**
     * List to display the results
     */
    private ListView lvPersons;
    private RecentsManager recentsManager;

    public PersonsSearchActivity() {
        super(TUMOnlineConst.Companion.getPERSON_SEARCH(), R.layout.activity_persons, PersonSearchSuggestionProvider.AUTHORITY, 3);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        lvPersons = findViewById(R.id.lstPersons);
        lvPersons.setOnItemClickListener(this);

        // get all stations from db
        recentsManager = new RecentsManager(this, RecentsManager.PERSONS);

        // Initialize persons adapter
        ArrayList<Person> list;
        ArrayAdapter<Person> adapter;
        try (Cursor personsCursor = recentsManager.getAllFromDb()) {
            list = new ArrayList<>(personsCursor.getCount());
            if (personsCursor.moveToFirst()) {
                do {
                    String recent = personsCursor.getString(0);
                    String[] t = recent.split("\\$");
                    Person p = new Person();
                    p.setId(t[0]);
                    p.setName(t[1]);
                    list.add(p);
                } while (personsCursor.moveToNext());
            }
        }

        adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, android.R.id.text1, list);

        if (adapter.getCount() == 0) {
            openSearch();
        } else {
            lvPersons.setAdapter(adapter);
            lvPersons.setOnItemClickListener(this);
            lvPersons.requestFocus();
        }
    }

    @Override
    public void onItemClick(AdapterView<?> a, View v, int position, long id) {
        Person person = (Person) lvPersons.getItemAtPosition(position);

        // store selected person ID in bundle to get in in StaffDetails
        Bundle bundle = new Bundle();
        bundle.putSerializable("personObject", person);

        // show detailed information in new activity
        Intent intent = new Intent(this, PersonsDetailsActivity.class);
        intent.putExtras(bundle);
        startActivity(intent);

        recentsManager.replaceIntoDb(person.getId() + "$" + person.toString()
                                                                  .trim());
    }

    @Override
    protected void onStartSearch() {
        ArrayList<Person> list;
        try (Cursor personsCursor = recentsManager.getAllFromDb()) {
            if (personsCursor.getCount() == 0) {
                finish();
                return;
            }

            list = new ArrayList<>(personsCursor.getCount());
            if (personsCursor.moveToFirst()) {
                do {
                    String recent = personsCursor.getString(0);
                    String[] t = recent.split("\\$");
                    Person p = new Person();
                    p.setId(t[0]);
                    p.setName(t[1]);
                    list.add(p);
                } while (personsCursor.moveToNext());
            }
        }
        ListAdapter adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, android.R.id.text1, list);
        lvPersons.setAdapter(adapter);
    }

    @Override
    public void onStartSearch(String query) {
        requestHandler.setParameter(P_SUCHE, query);
        requestFetch();
    }

    private void proceedToPersonDetails(PersonList response )    {
        lvPersons.setAdapter(null);
        Bundle bundle = new Bundle();
        bundle.putSerializable("personObject", response.getPersons().get(0));
        Intent intent = new Intent(this, PersonsDetailsActivity.class);
        intent.putExtras(bundle);
        startActivity(intent);
    }

    /**
     * Handles the XML response from TUMOnline by de-serializing the information
     * to model entities.
     *
     * @param response The de-serialized data from TUMOnline.
     */
    @Override
    public void onLoadFinished(PersonList response) {
        if (response.getPersons() == null || response.getPersons().isEmpty()) {
            lvPersons.setAdapter(new NoResultsAdapter(this));
        } else if (response.getPersons().size() == 1){
            proceedToPersonDetails(response);
        } else {
            ListAdapter adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, android.R.id.text1, response.getPersons());
            lvPersons.setAdapter(adapter);
        }
    }
}

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/person/PersonsSearchActivity.kt
package de.tum.`in`.tumcampusapp.component.tumui.person

import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.api.tumonline.TUMOnlineConst
import de.tum.`in`.tumcampusapp.component.other.general.RecentsDao
import de.tum.`in`.tumcampusapp.component.other.general.model.Recent
import de.tum.`in`.tumcampusapp.component.other.generic.activity.ActivityForSearchingTumOnline
import de.tum.`in`.tumcampusapp.component.tumui.person.model.Person
import de.tum.`in`.tumcampusapp.component.tumui.person.model.PersonList
import de.tum.`in`.tumcampusapp.database.TcaDb
import kotlinx.android.synthetic.main.activity_persons.*

/**
 * Activity to search for employees.
 */
class PersonsSearchActivity : ActivityForSearchingTumOnline<PersonList>(
        TUMOnlineConst.PERSON_SEARCH,
        R.layout.activity_persons,
        PersonSearchSuggestionProvider.AUTHORITY, 3
), PersonSearchResultsItemListener {

    private val recentsDao = TcaDb.getInstance(this).recentsDao()

    private val recents: List<Person>
        get() {
            val recents = recentsDao.getAll(RecentsDao.PERSONS) ?: return emptyList()
            return recents.map { recent -> Person.fromRecent(recent) }
        }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val layoutManager = LinearLayoutManager(this)

        personsRecyclerView.setHasFixedSize(true)
        personsRecyclerView.layoutManager = layoutManager

        val adapter = PersonSearchResultsAdapter(recents, this)
        if (adapter.itemCount == 0) {
            openSearch()
        }
        personsRecyclerView.adapter = adapter

        val itemDecoration = DividerItemDecoration(this, layoutManager.orientation)
        personsRecyclerView.addItemDecoration(itemDecoration)
    }

    override fun onItemSelected(person: Person) {
        val lastSearch = person.id + "$" + person.getFullName().trim { it <= ' ' }
        recentsDao.insert(Recent(lastSearch, RecentsDao.PERSONS))
        showPersonDetails(person)
    }

    override fun onStartSearch() {
        recentsHeader.visibility = View.VISIBLE
        val adapter = personsRecyclerView.adapter as? PersonSearchResultsAdapter
        adapter?.update(recents)
    }

    public override fun onStartSearch(query: String) {
        requestHandler.setParameter(P_SUCHE, query)
        requestFetch()
    }

    private fun proceedToPersonDetails(person: Person) {
        personsRecyclerView.adapter = null
        showPersonDetails(person)
    }

    private fun showPersonDetails(person: Person) {
        // Store selected person ID in bundle to get in in StaffDetails
        val bundle = Bundle().apply {
            putSerializable("personObject", person)
        }

        // Show detailed information in new activity
        val intent = Intent(this, PersonsDetailsActivity::class.java).apply {
            putExtras(bundle)
        }
        startActivity(intent)
    }

    /**
     * Handles the XML response from TUMOnline by de-serializing the information
     * to model entities.
     *
     * @param response The de-serialized data from TUMOnline.
     */
    public override fun onLoadFinished(response: PersonList) {
        recentsHeader.visibility = View.GONE

        if (response.persons.size == 1) {
            proceedToPersonDetails(response.persons[0])
        } else {
            val adapter = personsRecyclerView.adapter as? PersonSearchResultsAdapter
            adapter?.update(response.persons)
        }
    }

    companion object {
        private const val P_SUCHE = "pSuche"
    }

}
